feat: add cloud id during native sync

shenlong-tanwen opened the pull request:
## Description

- adds a new `cloudId` column to the local asset entity
- during the local sync, on iOS, the cloudId <-> localId mapping is fetched and updated to sqlite
- - the beta sync settings has a sync cloud ids button that will update the cloud ids of the already uploaded assets to the asset metadata table on the server
- The asset metadata on the server is updated back to the app to a new remote asset -> cloud id mapping table
- During hashing, if the local asset has a remote asset with a matching cloud ID, it'll reuse the checksum rather than calculating it on the device

## Note

- Need to confirm if editing an asset updates the cloud id
- From my testing, even local on device assets have a cloud id on iOS. Need to test if two Apple devices using the same account have consistent cloud ids for on device assets


Fixes #20360

FawenYo commented:
Hi, quick question about this PR:

My family and I currently use iCloud's Shared Photo Library to share some of our photos, and we're now planning to use Immich as well, while each of us has a separate Immich account.
My question is: if my account/device has already completed the hashing process, will the changes in this PR share the cloudId information with other Immich users? This would allow them to avoid spending a significant amount of time rehashing the same photos.

zackpollard commented:
> Hi, quick question about this PR:
> 
> My family and I currently use iCloud's Shared Photo Library to share some of our photos, and we're now planning to use Immich as well, while each of us has a separate Immich account. My question is: if my account/device has already completed the hashing process, will the changes in this PR share the cloudId information with other Immich users? This would allow them to avoid spending a significant amount of time rehashing the same photos.

It will not be shared between users if someone else has uploaded them previously, however if they have hashed them previously we may eventually add a mechanism for the client to store IDs for assets they don't upload, though we haven't decided what that would look like yet. This would save the devices from future hashing of assets on the device that you don't want to upload.

jrasm91 commented:
@shenlong-tanwen what's the status of this PR?

shenlong-tanwen committed:
use adjustment time in iOS for hash reset

# Conflicts:
#	mobile/lib/infrastructure/repositories/local_album.repository.dart
#	mobile/lib/presentation/pages/drift_asset_troubleshoot.page.dart

shenlong-tanwen committed:
migration

shenlong-tanwen committed:
feat: sync cloudId and eTag on sync

shenlong-tanwen committed:
fixes fixes

shenlong-tanwen committed:
more fixes

shenlong-tanwen committed:
re-sync updated eTags

shenlong-tanwen committed:
add server version check & auto sync cloud ids on compatible servers

shenlong-tanwen committed:
fix test

shenlong-tanwen committed:
remove button from sync status page

alextran1502 committed:
chore: modify for testing

shenlong-tanwen committed:
more changes

shenlong-tanwen committed:
chore: add commas in toString

alextran1502 committed:
Merge branch 'main' into feat/sync-cloud-id

shenlong-tanwen committed:
use cached provider in splash screen

shenlong-tanwen committed:
read upload service provider to prevent reset

shenlong-tanwen committed:
log errors from fetching cloud id mapping

shenlong-tanwen committed:
WIP: migrate cloud id - debug log

shenlong-tanwen committed:
ignore locked asset update

shenlong-tanwen committed:
merge main

# Conflicts:
#	mobile/lib/infrastructure/repositories/local_asset.repository.dart

shenlong-tanwen committed:
bulk update metadata

shenlong-tanwen committed:
merge main

alextran1502 committed:
Merge branch 'main' into feat/sync-cloud-id

shenlong-tanwen committed:
change log text
